### PR TITLE
Add intersection between Quadrangles and rays/segments

### PR DESCRIPTION
**Pull request by HaoxuanGuo**
An extra step towards #125 . I tried to write `[intersect(ray, triangle) for triangle in mesh]` but it failed because there were some `Quadrangle`s. So, this PR is to fix that.

### CODE REVIEW

**Comment by GuoHaoxuan** [comment]
Do not merge this! Here is a bug when the `Ngon` is a concave polygon. Sorry about that.

**Comment by GuoHaoxuan** [comment]
At least `Quadrangle` should be considered. I wrote this condition and considered concave quadrangle.

File: src/intersections/quadrangle.jl
Comment (juliohm) [inline src/intersections/quadrangle.jl:12]: I think we need to organize this better. We don't want to write it in terms of a Union type. Also, I think the internal implementation should be expressed in terms of intersecttype so that the function `f` can be forwarded. More generally, I think we should consider and `Ngon` as input, triangulate this ngon with the existing triangulation methods, and check the intersection.

**Review by juliohm — CHANGES_REQUESTED** [review]
Hi @HaoxuanGuo I think we need to write this down more carefully and more generically in terms of `Polygon` instead. Given a polygon, you can triangulate it, perform intersections and return the appropriate value. This could even be done in parallel with multiple threads in the sense that each triangle making up the polygon is checked indenpendently in a separate thread.

**Comment by GuoHaoxuan** [comment]
Close due to #214 .

**Pull request closed without merging**
